### PR TITLE
Secd 782/handle 0timestamp assets

### PR DESCRIPTION
**Pull request by realBrandonChan**
This branch is intended to apprehend bad Asset histories, specifically those that are labeled as "SCAN" but have a blank timestamp. If a timestamp is blank, it automatically gets initialized with a 0 value time, thus we need to have a stricter check.

### CODE REVIEW

**Comment by bkrebsbach** [comment]
The build is failing.

File: pkg/assetfetcher/asset.go [hunk 1]
```diff
@@ -339,7 +339,8 @@ func (a assetHistoryEvents) lastScannedTimestamp() (time.Time, error) {
 			if err != nil {
 				return latestTime, err
 			}
-			if t.After(latestTime) {
+			// also need to make sure the date is not a 0 value
+			if t.After(latestTime) && !t.IsZero() {
```
Comment (bkrebsbach) [inline pkg/assetfetcher/asset.go]: Task: This check is unnecessary. At this point, `latestTime` will either be: 1) the initial zero value, so `t.After(latestTime)` would be false if t is also a zero value, or 2) `latestTime` will be non-zero, and t.After will still return false is t is the zero value.

File: pkg/assetfetcher/asset_test.go [hunk 2]
```diff
@@ -47,6 +47,30 @@ func TestLastAssessedForVulnerabilities(t *testing.T) {
...
+			time.Time{},
+			true,
+		},
+		{
+			"0 timestamp field",
+			assetHistoryEvents{
+				AssetHistory{Type: "SCAN", Date: "0001-01-01T00:00:00Z"},
+			},
+			time.Time{},
+			false,
```
Comment (bkrebsbach) [inline pkg/assetfetcher/asset_test.go]: Question: shouldn't we expect an error with a zero value timestamp?

File: pkg/assetfetcher/asset.go (hunk 1 above)
Comment (realBrandonChan) [reply pkg/assetfetcher/asset.go]: yes! you are indeed correct, good point. Fixed!

File: pkg/assetfetcher/http.go [hunk 3]
```diff
@@ -183,7 +183,13 @@ func (c *NexposeAssetFetcher) newNexposeSiteAssetsRequest(siteID string, page in
 func (a Asset) hasBeenScanned() bool {
 	for _, evt := range a.History {
```
Comment (bkrebsbach) [inline pkg/assetfetcher/http.go]: Question: aren't we walking the asset history in `AssetPayloadToAssetEvent()` and here?

File: pkg/assetfetcher/asset_test.go (hunk 2 above)
Comment (realBrandonChan) [reply pkg/assetfetcher/asset_test.go]: Good question, lastScannedTimestamp() won't return an error with this case, it will return the zero value timestamp. However, the only method that calls lastScannedTimestamp() is AssetPayloadToAssetEvent(), and that checks if lastScannedTimestamp() returns a zero value, which will then result in an error.

**Comment by bkrebsbach** [comment]
Question: have you confirmed that we're seeing SCAN histories with blank/invalid timestamps? 

Looking at the original issue and the code, I believe the reason we see the empty timestamp errors is because we're duplicating the asset check in pkg/assetfetcher/http.go:
* On [line 104](https://github.com/asecurityteam/nexpose-asset-producer/pull/26/files#diff-870e38936ad1e76d11e27c416083303cL104), for the first call to Nexpose we're stating assets that haven't been scanned and skipping them.
* All subsequent calls use makeRequest - [line 121](https://github.com/asecurityteam/nexpose-asset-producer/pull/26/files#diff-870e38936ad1e76d11e27c416083303cR121)
* On [line 160](https://github.com/asecurityteam/nexpose-asset-producer/pull/26/files#diff-870e38936ad1e76d11e27c416083303cR160) you can see we're not making the same check

File: pkg/assetfetcher/http.go [hunk 4]
```diff
@@ -183,7 +189,13 @@ func (c *NexposeAssetFetcher) newNexposeSiteAssetsRequest(siteID string, page in
...
 		if evt.Type == "SCAN" {
-			return true
+			t, err := time.Parse(time.RFC3339, evt.Date)
+			// check if the date field is parsable and if it isn't 0 value
+			if err != nil || t.IsZero() {
+				continue
+			} else {
```
Comment (bkrebsbach) [inline pkg/assetfetcher/http.go]: Suggestion: `else` is not really considered idiomatic in go - https://golang.org/doc/effective_go.html#if. Since you have a continue on line 195, you can drop the else and just have a bare `return true` after the if block.

File: pkg/assetfetcher/http_test.go [hunk 5]
```diff
@@ -545,7 +546,7 @@ func TestFetchAssetsAssetPayloadToAssetEventError(t *testing.T) {
 	mockRT := NewMockRoundTripper(ctrl)
 
 	resp := SiteAssetsResponse{
-		Resources: []Asset{{History: assetHistoryEvents{AssetHistory{Type: "SCAN", Date: "not a time"}}}},
+		Resources: []Asset{{History: assetHistoryEvents{AssetHistory{Type: "SCAN", Date: "2019-05-14T15:03:47.000Z"}}}},
```
Comment (bkrebsbach) [inline pkg/assetfetcher/http_test.go]: Question: I'm not seeing why this change was made. Why does this asset fail to convert?

File: pkg/assetfetcher/http_test.go [hunk 6]
```diff
@@ -700,7 +702,7 @@ func TestMakeRequestWithAssetPayloadToAssetEventError(t *testing.T) {
 	defer ctrl.Finish()
 	mockRT := NewMockRoundTripper(ctrl)
 	resp := SiteAssetsResponse{
-		Resources: []Asset{{History: assetHistoryEvents{AssetHistory{Type: "SCAN", Date: "not a time"}}}},
+		Resources: []Asset{{History: assetHistoryEvents{AssetHistory{Type: "SCAN", Date: "1234-12-01T00:00:00Z"}}}},
```
Comment (bkrebsbach) [inline pkg/assetfetcher/http_test.go]: Question: same as above, why is this change necessary?

File: pkg/assetfetcher/http_test.go [hunk 7]
```diff
@@ -850,3 +852,19 @@ func TestNewNexposeSiteAssetsRequestWithExtraSlashes(t *testing.T) {
 
 	assert.Equal(t, "http://localhost/api/3/sites/siteID/assets?page=1&size=100", req.URL.String())
 }
+func TestHasBeenScannedBadTimeField(t *testing.T) {
+	dummyAssetBlankTimeStamp := Asset{History: assetHistoryEvents{AssetHistory{Type: "SCAN", Date: ""}}}
```
Comment (bkrebsbach) [inline pkg/assetfetcher/http_test.go]: Suggestion: if you're going to combine multiple test cases in a single test function, you should use [table driven tests](https://github.com/golang/go/wiki/TableDrivenTests) to make it more clear what your test cases are vs your test harness.

This is a good example within this code base: https://github.com/asecurityteam/nexpose-asset-producer/blob/master/pkg/assetfetcher/errors_test.go#L75

File: pkg/assetfetcher/asset_test.go [hunk 8]
```diff
@@ -106,36 +114,29 @@ func TestAssetPayloadToAssetEventError(t *testing.T) {
...
+// check if we do indeed receive a asset history with an invalid history
+// AssetPayloadToAssetEvent never should receive an invalid asset history,
+// this test is more so to have coverage, and make sure we do catch this in the event
+func TestAssetPayloadToAssetEventLastScannedError(t *testing.T) {
 	tests := []struct {
 		name  string
 		asset Asset
 	}{
 		{
-			"No LastScanned",
```
Comment (bkrebsbach) [inline pkg/assetfetcher/asset_test.go]: Suggestion: I think it's a good idea to leave this test case in, just to confirm the behavior of this function if an asset has no history.

File: pkg/assetfetcher/asset.go [hunk 9]
```diff
@@ -313,12 +313,14 @@ type VulnerabilitySummary struct {
 // into an AssetEvent for downstream services.  Callers of this function are
 // expected to pass an Asset that cleanly maps to AssetEvent, meaning the
 // event must have a timestamp (because without one, it's not an event!)
+// It is guaranteed that when this function is called an asset will have a valid
```
Comment (bkrebsbach) [inline pkg/assetfetcher/asset.go]: Nit: guaranteed is a strong word here. We could very easily change something in pkg/assetfetcher/http.go that invalidates this statement. I think it would be more correct to say something like "This function expects an asset will have a valid scanned history"

File: pkg/assetfetcher/http_test.go (hunk 6 above)
Comment (realBrandonChan) [reply pkg/assetfetcher/http_test.go]: I added a check before AssetPayloadtoAssetEvent() to validate the asset history, so in order to trigger AssetPayloadtoAssetEvent(), I needed to change the Date field to be something valid.

File: pkg/assetfetcher/http_test.go (hunk 5 above)
Comment (realBrandonChan) [reply pkg/assetfetcher/http_test.go]: I added a check before AssetPayloadtoAssetEvent() to validate the asset history, so in order to trigger AssetPayloadtoAssetEvent(), I needed to change the Date field to be something valid.

File: pkg/assetfetcher/http.go (hunk 4 above)
Comment (realBrandonChan) [reply pkg/assetfetcher/http.go]: done!

File: pkg/assetfetcher/asset.go (hunk 9 above)
Comment (realBrandonChan) [reply pkg/assetfetcher/asset.go]: fair point. done!

File: pkg/assetfetcher/http_test.go (hunk 7 above)
Comment (realBrandonChan) [reply pkg/assetfetcher/http_test.go]: can do
